backend/manager/modules/bll/src/main/java/org/ovirt/engine/core/bll/storage/disk/image/ConvertDiskCommand.java

```diff
@@ -0,0 +1,333 @@
+        if (getDiskImage() == null) {
+            DiskImage diskImage = diskImageDao.get(locationInfo.getImageId());
+            setDiskImage(diskImage);
+        }
+    }
+
+    @Override
+    protected boolean validate() {
+        StorageDomain storageDomain = storageDomainDao.get(getDiskImage().getStorageIds().get(0));
+        if (storageDomain == null) {
+            return failValidation(EngineMessage.STORAGE_DOMAIN_DOES_NOT_EXIST);
+        }
+
+        // We do not support chains currently
+        if (diskImageDao.getAllSnapshotsForImageGroup(getDiskImage().getId()).stream().count() > 1) {
+            return failValidation(EngineMessage.ACTION_TYPE_FAILED_DISK_WITH_CHAIN);
+        }
+
+        StorageDomainValidator storageDomainValidator = new StorageDomainValidator(storageDomain);
+        ValidationResult result = storageDomainValidator.hasSpaceForNewDisk(getDiskImage());
+        if (!result.isValid()) {
+            failValidation(result.getMessages());
+        }
+
+        // We do not support live stuff currently
+        Map<Boolean, List<VM>> vms = vmDao.getForDisk(getDiskImage().getId(), true);
+
+        boolean runningVms = vms.computeIfAbsent(Boolean.TRUE, b -> new ArrayList<>())
+                .stream()
+                .anyMatch(vm -> vm.isRunningOrPaused());
+        if (!vms.isEmpty() && runningVms) {
+            return failValidation(EngineMessage.ACTION_TYPE_FAILED_VM_IS_RUNNING);
+        }
+
+        // Make sure unsupported combinations aren't used (sparse-raw on block, etc)
+        return ImagesHandler.checkImageConfiguration(storageDomain.getStorageStaticData(),
+                getVolumeType(getDiskImage()),
+                getVolumeFormat(getDiskImage()),
+                getDiskImage().getBackup(),
+                getReturnValue().getValidationMessages());
+    }
+
+    @Override
+    protected void executeCommand() {
+        Guid hostForExecution = vdsCommandsHelper.getHostForExecution(getDiskImage().getStoragePoolId(),
+                FeatureSupported::isConvertDiskSupported);
+        if (Guid.isNullOrEmpty(hostForExecution)) {
+            log.error("Could not find host to perform conversion");
+            return;
+        }
+
+        getParameters().setVdsRunningOn(hostForExecution);
+
+        // Disable pre-polling nonsense
+        getParameters().setEntityInfo(new EntityInfo(VdcObjectType.Disk, getDiskImage().getId()));
+        getParameters().setNewVolGuid(Guid.newGuid());
+        // Create volume in the same disk
+        runInternalAction(ActionType.CreateVolumeContainer, createVolumeCreationParameters(getDiskImage()));
+        getParameters().setConvertDiskPhase(ConvertDiskCommandParameters.ConvertDiskPhase.CONVERT_VOLUME);
+
+        persistCommandIfNeeded();
+        setSucceeded(true);
+    }
+
+    @Override
+    public boolean performNextOperation(int completedChildCount) {
+        persistCommandIfNeeded();
```

Comment:
don't we need to persist the command after setting its next phase?

Reply:
moved to `updatePhase`